docs(replay): Replay Issue types & hydration errors (#10686)
getsentry/sentry-docs: docs/product/issues/issue-details/replay-issues/hydration-error.mdx

---
title: Hydration Error
sidebar_order: 50

> description: "Learn about hydration errors."

Reviewer: ```suggestion
description: "Learn about Session Replay hydration errors."
```

Reviewer: to match the description on https://docs.sentry.io/product/issues/issue-details/replay-issues/

---

> Hydration errors are a React-specific problem that happen when the initial UI does not match what was rendered on the server. They can result in extra work for the browser, and a slower pageload experience for users.

Reviewer 2: wondering if we should stick the word `client` in here somewhere to reinforce what is not matching against server

```suggestion
Hydration errors are a React-specific problem that happen when the initial client UI does not match what was rendered on the server. They can result in extra work for the browser, and a slower pageload experience for users.
```

## Prerequisites for Seeing Hydration Error Diff Tools

> Hydration Errors can come into your issue stream either as regular error events, or they could be detected inside a replay session and converted into an issue by the server. Read more about

Reviewer: ```suggestion
Hydration Errors can come into your issue stream either as regular error events, or they could be detected inside a replay and converted into an issue by the server. Read more about
```

Reviewer: This is really confusing :/ Can we clarify that hydration errors from 'regular error events' are filtered out via inbound filters

Author: The linked page `[configuring hydration errors](/platforms/javascript/session-replay/issue-types/)` does talk about the inbound filters and what the different preference options are, i think it's ok to only have that in one spot. Also, it is confusing!

> <PlatformLink to="/session-replay/issue-types/">configuring hydration errors</PlatformLink> to learn about the differences.

Reviewer 3: ```suggestion
[configuring hydration errors](https://docs.sentry.io/platforms/javascript/session-replay/issue-types/)o learn about the differences.
```

Changing this to a direct link so that it doesn't prompt you to choose JS when that is the only option

> If the hydration error event instance has an associated replay session, then you will see diff tools on the issue details page to make debugging the hydration error easier.

Reviewer: ```suggestion
If the hydration error event instance has an associated replay, then you will see diff tools on the **Issue Details** page to make debugging the hydration error easier.
```

Reviewer: isn't this always the case since these issues are session replay-based? i think this paragraph/section needs to be re-thought.

we should start from the standpoint of 'session replay specific hydration errors' and talk about what those entail.

then, in a small callout box we can reference the traditional hydration error issues and how those are different and filtered out by default etc

Author: both replay-hydration and error-hydration get the same set of diff-tools if a replay is attached. It just turns out that in the former case a replay is guaranteed to be there. How the issue got into the stream is a bit of an implementation problem though, and probably we should only talk about it in terms of the preferences (which people shouldn't need to touch imo).

I don't think we should talk too much about "traditional hydration error" being very different, because when you're debugging the issue it's not really different, the page looks the same.

Reviewer: @ryan953  🫨 oooh i didn't know that replay-hydration AND error-hydration both get the new and improved diffing treatment.

i still find referencing both kind of issues here confusing and its not clear what each one entails/why it matters

i think it makes sense to focus on replay-hydration since that's this page is dedicated on, and the other traditional hydration error is more an 'aside'/callout

Author: maybe the page needs to be just about hydration errors in general, and the inbound-filters and replay-project-settings stuff can be somewhere less prominent (therefore less confusing)

## Debugging Hydration Errors

Hydration errors represent a mismatch between the server-rendered HTML, and the client-rendered HTML. So it's beneficial to compare the before and after state of a webpage to understand what's changed, and decide how to make those two states consistent. To do this, we provide a few diff tools to see the differences -- both visually and by looking at the rendered HTML itself.

> Generally the visual tools are a good place to start to identify the issue, but keep in mind that sometimes differences can appear below the fold, or there can be differences in hidden DOM nodes or attributes, so inspecting the HTML is important too.

Reviewer: ```suggestion
Generally, the visual tools are a good place to start to identify the issue, but keep in mind that sometimes differences can appear below the fold, or there can be differences in hidden DOM nodes or attributes, so inspecting the HTML is important too.
```

### Example Error

In the example below we'll look at a hydration error and see how each tool helps to debug the problem.

The problem is that code highlighting is not applied on the server. When the server renders the PHP code block in the bottom-center of the page, all the code has a white font color. After hydration syntax highlighting is applied, the code snippet is colorful.

#### Image Slider Tool

The image slider tool allows for comparing the two page states by overlaying the images on top of each other. Click and drag the purple line to reveal one image or the other.

> ![image slider tool](./img/hydration-error-image-slider-tool.png)

Reviewer 3: This and the next images are so tiny it's really hard to even tell what they are showing. Can we make them a bit more zoomed in?

Reviewer 3: Oh awesome :D

> #### Side by Side Image Tool

Reviewer: ```suggestion
#### Side-by-Side Image Tool
```

> View images side by side to see large chunks of the page that may differ.

Reviewer: ```suggestion
View images side-by-side to see large chunks of the page that may differ.
```

> ![side by side image tool](./img/hydration-error-side-by-side-image-tool.png)

Reviewer 3: This is the other image that is really hard to read

Author: higher res incoming!

#### HTML Diff Tool

> The HTML diff tool is a basic HTML comparison. It is useful if the hydration error is caused by hidden DOM nodes, different attributes on a DOM node, or mismatched content below the fold.

Reviewer: ```suggestion
The HTML diff tool is a basic HTML comparison. It is useful if the hydration error is caused by hidden DOM nodes, different attributes on a DOM node, or there's mismatched content below the fold.
```

![html diff tool](./img/hydration-error-html-diff-tool.png)

## Privacy

> The diff tools all rely on replay data, so any PII data within the webpage will be masked or blocked before being sent to the server. Read more about [Protecting user privacy in Session Replay](/security-legal-pii/scrubbing/protecting-user-privacy/) and [how to configure Session Replay to maintain user privacy](/platforms/javascript/session-replay/privacy/).

Reviewer: ```suggestion
The diff tools all rely on replay data, so any PII data within the webpage will be masked or blocked before being sent to the server by default. Read more about [Protecting user privacy in Session Replay](/security-legal-pii/scrubbing/protecting-user-privacy/) and [how to configure Session Replay to maintain user privacy](/platforms/javascript/session-replay/privacy/).
```

> The diff tools all rely on replay data, so any PII data within the webpage will be masked or blocked before being sent to the server. Read more about [Protecting user privacy in Session Replay](https://docs.sentry.io/security-legal-pii/scrubbing/protecting-user-privacy/) and <PlatformLink to="/session-replay/privacy/">how to configure Session Replay to maintain user privacy</PlatformLink>.

Reviewer 3: ```suggestion
The diff tools all rely on replay data, so any PII data within the webpage will be masked or blocked before being sent to the server. Read more about [Protecting user privacy in Session Replay](https://docs.sentry.io/security-legal-pii/scrubbing/protecting-user-privacy/) and [how to configure Session Replay to maintain user privacy](https://docs.sentry.io/platforms/javascript/session-replay/privacy/).
```

Again, replacing with a link so that they don't have to click JS when it's the only option
